sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/SonarAnalysisContextTest.cs

```diff
@@ -130,5 +107,66 @@ public void SonarAnalysis_IssuesAreReportedByDefault()
 
             shouldIssueBeReported(null, dummyDiag).Should().BeTrue();
         }
+
+        [TestMethod]
+        public void SonarAnalysis_WhenOneRuleDisabled_ReportIssuesOnlyForOtherRules()
+        {
+            try
+            {
+                SonarAnalysisContext.ShouldExecuteRuleFunc = context => context.SupportedDiagnostics
+                    .Any(diagnostic => diagnostic.Id != AnonymousDelegateEventUnsubscribe.DiagnosticId);
+
+                foreach (var testCase in TestCases)
+                {
+                    if (testCase.Analyzer.GetType() == typeof(AnonymousDelegateEventUnsubscribe))
+                    {
+                        Verifier.VerifyNoIssueReported(testCase.Path, testCase.Analyzer);
+                    }
+                    else
+                    {
+                        Verifier.VerifyAnalyzer(testCase.Path, testCase.Analyzer);
+                    }
+                }
+            }
+            finally
+            {
+                SonarAnalysisContext.ShouldAnalysisBeDisabled = null;
+            }
+        }
+
+        [TestMethod]
+        public void SonarAnalysis_WhenReportDiagnosticActionNotNull_AllowToContolWhetherOrNotToReport()
+        {
+            try
+            {
+                SonarAnalysisContext.ReportDiagnosticAction = context =>
+                {
+                    if (context.Diagnostic.Id != AnonymousDelegateEventUnsubscribe.DiagnosticId)
+                    {
+                        // Verifier expects all diagnostics to increase the counter in order to check that all rules call the
+                        // extension method and not the direct `ReportDiagnostic`.
+                        Verifier.counters.AddOrUpdate(context.Diagnostic.Id,
```

Comment:
Changing the dictionary directly seems a bit hacky. How about adding an explicit method to register that the analyser has been called, or is it possible to modify _HookSuppression_ to handle the new delegate too?

Reply:
The `Verifier` class got really complex over time and really need a good refactoring. It is nearly impossible to extend it and really hard to add some extra checks.

Reply:
Ok. Adding a method _record that analyzer reported_ method would be better than directly changing the collection, though.